Support redirection in client connector

arukshani opened the pull request:
## Purpose
> Support http redirection in http client connector. 

## Goals
> When a redirect response is received, before sending it to the client follow the redirect and get it's response. Maximum of 5 redirects are allowed by default. 

## Approach
> Add a redirect handler to the pipeline and if the redirect request goes to the same domain, use the existing channel to send the request, otherwise create a new channel for the redirect request.
[https://docs.google.com/a/wso2.com/document/d/1Mf89vZ5c43eVx9p3rQhtsVHDiMDrxuqHfxVIgzJC48c/edit?usp=sharing](url)

## Automation tests
> Test coverage -  82%

Commit:
Merge branch 'master' of https://github.com/wso2/carbon-transports into redirect-properties

Commit:
Intial code for redirection support for client connector

Commit:
Merge branch 'master' of https://github.com/wso2/carbon-transports into redirect-properties

Commit:
Add test cases for redirect handling in client connector

Commit:
Merge branch 'master' of https://github.com/wso2/carbon-transports into redirect-properties

Commit:
Set redirect request's header to include host

Commit:
Merge branch 'master' of https://github.com/wso2/carbon-transports into redirect-properties

Commit:
Add unit test case for redirect loop test
and set the request as an attribute to the inbound channel

Commit:
Get the default port based on the http protocol in RedirectHandler

Commit:
When the original channel is send the response back to the client directly through redirect handler.

Commit:
Add unit test cases for each redirect status codes
Unit test cases to test whether the redirection happens using the correct http method, according to the redirect status code and
check whether the cross domains urls are handled properly.

Commit:
Set the timeout for redirect channel.
Timeout for the redirect channel has been calculated as follows.
 long timeElapsedSinceOriginalRequest = currentTime - channelStartTime;
 long remainingTimeForRedirection = timeoutOfOriginalRequest - timeElapsedSinceOriginalRequest;

Commit:
Merge branch 'master' of https://github.com/wso2/carbon-transports into redirect-properties

Commit:
Pass original timeout and channel start time down each child channel
When a idle handle is added to the channel, remove the previous channel's idle handler

Commit:
Add integration test case for redirect timeout test
Change the TestServer to accept a delayed timout in milliseconds.

pamod commented on http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/common/Util.java:
```diff
@@ -494,4 +497,25 @@ public static String getSystemVariableValue(String variableName, String defaultV
         return host + ":" + port;
     }
+
+    public static HTTPCarbonMessage createErrorMessage(String payload) {
```
Shall we include javaDocs explaining this ?

arukshani replied:
done

pamod commented on http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/sender/RedirectChannelInitializer.java:
```diff
@@ -0,0 +1,81 @@
+
+    private static final Logger log = LoggerFactory.getLogger(HTTPClientInitializer.class);
+
+    private SSLEngine sslEngine;
```
It would look more precise if we could explain the role of each variable ?

arukshani replied:
done

pamod commented on http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/sender/RedirectChannelInitializer.java:
```diff
@@ -0,0 +1,81 @@
+        // Add the generic handlers to the pipeline
+        // e.g. SSL handler
+        if (sslEngine != null) {
+            log.debug("adding ssl handler");
```
Shall we wrap this with isDebugEnabled()

arukshani replied:
done

pamod commented on http/org.wso2.carbon.transport.http.netty/src/test/java/org/wso2/carbon/transport/http/netty/util/TestUtil.java:
```diff
@@ -175,6 +175,24 @@ public static HttpsServer startHttpsServer(int port, String message, String cont
+            httpServer.setLocation(location);
+            httpServer.setDelay(delay);
+        } catch (Exception e) {
+            log.error("Thread Interrupted while sleeping ", e);
```
This will not only grab interrupted exceptions ? any reason for using generic exceptions here ?

arukshani replied:
Changed to capture only InterruptedException

indikasampath2000 commented on http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/sender/RedirectChannelInitializer.java:
```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (c) 2015, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Could you update license header year to 2017?

arukshani replied:
done

Commit:
Add missing javadocs.

Commit:
Maintain redirect count in each redirect channel instead of the original channel.
This is needed in case the original channel gets inactive, susequent channels will have no way to get the current redirect count.

Commit:
Merge branch 'master' of https://github.com/wso2/carbon-transports into redirect-properties

Commit:
Refactor RedirectHandler class
Add debug logs where necessary
Break larger functions into smaller functions

Commit:
Use same event loop for subsequent channels

Commit:
Merge branch 'master' of https://github.com/wso2/carbon-transports into redirect-properties

Commit:
Use the same eventloop group for newly created channels when redirecting
and use the original channel's future object always to notify response to the client

Commit:
Merge branch 'master' of https://github.com/wso2/carbon-transports into redirect-properties

Commit:
Return original channel back to its pool when the response is written to client

Commit:
Merge branch 'master' of https://github.com/wso2/carbon-transports into redirect-properties

Commit:
Add a flag to keep track of idle state handler removement from both the target channel  and redirect channel

Commit:
Fix idle handler duplicate issue when a timeout occurs in target channel.
Add unit test to check proper building of redirect url from a relative path.

Commit:
Correct error message for unit test case "requestPathEndsWithSlash"

Commit:
Set the default max redirect count in transport side as well, in case it didn't come from the ballerina side.

Commit:
In case of an exception in Redirecthandler call it's exceptionCaught method.

Commit:
When an exception is thrown in Redirecthandler, handle it using exceptionCaught method

Commit:
When the response is not a redirect response, let the target handler deal with the response.

Commit:
No need to keep track of idlehandler state for redirect channel as it gets closed once the message is returned
to the client

Commit:
When a timeout occured in the original channel, let the target hanlder handle it.

Commit:
Merge branch 'master' of https://github.com/wso2/carbon-transports into redirect-properties

shafreenAnfar commented on http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/common/Constants.java:
```diff
@@ -175,6 +180,14 @@
+
+    public static final String HTTP_POST_METHOD = "POST";
+
+    public static final String HTTP_HEAD_METHOD = "HEAD";
```
I think there is no need for space between the above lines as they are associated together. I am talking about the verbs.

arukshani replied:
Fixed this.

shafreenAnfar commented on http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/common/Util.java:
```diff
@@ -500,4 +504,42 @@ public static String getSystemVariableValue(String variableName, String defaultV
+        response.setHeader(Constants.HTTP_CONTENT_LENGTH, (String.valueOf(errorMessageBytes.length)));
+
+        response.setProperty(Constants.HTTP_STATUS_CODE, 504);
+        response.setProperty(org.wso2.carbon.messaging.Constants.DIRECTION,
```
This is not needed anymore.

arukshani replied:
Removed.

shafreenAnfar commented on http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/common/Util.java:
```diff
@@ -500,4 +504,42 @@ public static String getSystemVariableValue(String variableName, String defaultV
+     * @param payload Message payload.
+     * @return HTTPCarbonMessage
+     */
+    public static HTTPCarbonMessage createErrorMessage(String payload) {
```
Do we really need this method ?

arukshani replied:
Removed this method.

shafreenAnfar commented on http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/contractimpl/HttpClientConnectorImpl.java:
```diff
@@ -75,16 +82,25 @@ public HttpResponseFuture send(HTTPCarbonMessage httpCarbonRequest) {
                             if (isValidateChannel(channelFuture)) {
                                 targetChannel.setChannel(channelFuture.channel());
-
+                                channelFuture.channel().attr(Constants.ORIGINAL_REQUEST).set(httpCarbonRequest);
```
Do we need to do this for each and every request ?

arukshani replied:
Channel attributes are set only when the redirect is on.

shafreenAnfar commented on http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/sender/RedirectHandler.java:
```diff
@@ -0,0 +1,783 @@
+ */
+public class RedirectHandler extends ChannelInboundHandlerAdapter {
+
+    protected static final Logger LOG = LoggerFactory.getLogger(RedirectHandler.class);
```
Usually we have a space after this line.

Commit:
Remove error response message as it is not needed anymore

Commit:
Set channel properties only when the redirect is on.

prabushi commented on http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/common/Constants.java:
```diff
@@ -253,10 +264,36 @@
+            ("ORIGINAL_REQUEST");
+    public static final AttributeKey<HttpResponseFuture> RESPONSE_FUTURE_OF_ORIGINAL_CHANNEL = AttributeKey
+            .<HttpResponseFuture>valueOf
+            ("RESPONSE_FUTURE_OF_ORIGINAL_CHANNEL");
```
Shall we append this line with the above?

arukshani replied:
Fixed this. https://github.com/wso2/carbon-transports/pull/412/files#diff-9e2df4e3d298525a473d5d10e8c5d5e9R272

Bhashinee commented on http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/sender/RedirectChannelInitializer.java:
```diff
@@ -0,0 +1,77 @@
+        // e.g. SSL handler
+        if (sslEngine != null) {
+            if (log.isDebugEnabled()) {
+                log.debug("adding ssl handler");
```
Add a meaningful log here

arukshani replied:
Done.

prabushi commented on http/org.wso2.carbon.transport.http.netty/src/test/java/org/wso2/carbon/transport/http/netty/redirect/HTTPClientRedirectTestCase.java:
```diff
@@ -0,0 +1,565 @@
+        embeddedChannel.attr(Constants.REDIRECT_COUNT).set(5);
+        embeddedChannel.writeInbound(response);
+        embeddedChannel.writeInbound(LastHttpContent.EMPTY_LAST_CONTENT);
+        assertNull(embeddedChannel.readOutbound());
```
Shall we add the assert messages?

arukshani replied:
Done >> https://github.com/wso2/carbon-transports/pull/412/files#diff-44d72deb2696196bfa93f63a55a2069aL123

Asitha commented on http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/sender/RedirectHandler.java:
```diff
@@ -0,0 +1,782 @@
+                    }
+                } else {
+                    LOG.error("Error occurred while trying to connect to redirect channel.", future.cause());
+                    exceptionCaught(channelHandlerContext, future.cause());
```
Are we logging the exception twice?

arukshani replied:
This is fixed now. https://github.com/wso2/carbon-transports/pull/415

Asitha commented on http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/sender/RedirectHandler.java:
```diff
@@ -0,0 +1,782 @@
+            exceptionCaught(ctx, exception.getCause());
+        } catch (MalformedURLException exception) {
+            LOG.error("MalformedURLException occurred when deciding whether a redirection is required", exception);
+            exceptionCaught(ctx, exception.getCause());
```
Are we logging the exception twice? Shall we log it only once? Maybe we can pass the exception message as well to the exception caught method for logging

arukshani replied:
This is fixed now. Since the exceptionCaught is an overriden method from netty framework, to handle application logic errors, a separate method was defined. https://github.com/wso2/carbon-transports/pull/415

Asitha commented on http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/sender/RedirectHandler.java:
```diff
@@ -0,0 +1,782 @@
+                    "Error occurred while returning target channel " + targetChannel.getChannel().id() + " from current"
+                            + " channel" + ctx.channel().id() + " " + "to its pool in " + "markEndOfMessage",
+                    exception);
+            exceptionCaught(ctx, exception.getCause());
```
Are we logging the exception twice?

arukshani replied:
This is fixed in https://github.com/wso2/carbon-transports/pull/415
